presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java

```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+                    }
+                }
+            }
+        }
+
+        return unmodifiableList(stream(reduceTo)
+                .distinct()
+                .boxed()
+                .map(predicates::get)
+                .collect(toList()));
+    }
+
+    /**
+     * Eliminate a sub predicate if its sub predicates contain its peer. Will return null if cannot extract common predicates otherwise return a nested list with flipped form
+     * For example:
+     * (a || b || c || d) && (a || b || e || f)  -> a || b || ((c || d) && (e || f))
+     * (a || b) && (c || d) -> null
+     */
+    private List<List<RowExpression>> extractCommonPredicates(Form rootForm, List<List<RowExpression>> predicates)
+    {
+        if (predicates.isEmpty()) {
+            return null;
+        }
+        Set<RowExpression> commonPredicates = new LinkedHashSet<>(predicates.get(0));
+        predicates.forEach(commonPredicates::retainAll);
+
+        if (commonPredicates.isEmpty()) {
+            return null;
+        }
+        List<RowExpression> remainingPredicates = unmodifiableList(predicates
+                .stream()
+                .map(subPredicates ->
+                        combinePredicates(flip(rootForm),
+                                subPredicates
+                                        .stream()
+                                        .filter(predicate -> !commonPredicates.contains(predicate))
+                                        .collect(toList())))
+                .collect(toList()));
+        return Stream.concat(commonPredicates.stream().map(predicate -> singletonList(predicate)), Stream.of(remainingPredicates))
+                .collect(toList());
+    }
+
+    private RowExpression combineSubPredicates(Form clause, List<List<RowExpression>> predicates)
+    {
+        return combinePredicates(clause, predicates.stream()
+                .map(predicate -> combinePredicates(flip(clause), predicate))
+                .collect(toList()));
+    }
+
+    /**
+     * Cartesian cross product of List of List.
+     * For example, [[a], [b, c], [d]] becomes [[a,b,d], [a,c,d]]
+     */
+    private static List<List<RowExpression>> crossProduct(List<List<RowExpression>> lists)
+    {
+        checkArgument(lists.size() > 0, "Must contain more than one child");
+        List<List<RowExpression>> result = lists.get(0).stream().map(Collections::singletonList).collect(toList());
+        for (int i = 1; i < lists.size(); i++) {
+            result = crossProduct(result, lists.get(i));
+        }
+        return result;
+    }
+
+    private static List<List<RowExpression>> crossProduct(List<List<RowExpression>> left, List<RowExpression> right)
+    {
+        List<List<RowExpression>> result = new ArrayList<>();
+        for (List<RowExpression> combined : left) {
+            for (RowExpression toAdd : right) {
+                List<RowExpression> newLine = new ArrayList<>(combined);
```

Comment:
nit: `newClause`? 

Also maybe name `combined`  as `previousClause`, and `toAdd` as `newTerm`. So we combine `previousClause` and `newTerm` into `newClause`